### PR TITLE
Implement Copy, Deref and FromStr to (ed25519|sr25519)::Public

### PR DESCRIPTION
**Pull request by c410-f3r**

### CODE REVIEW

File: core/primitives/src/ed25519.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ use crate::crypto::{Pair as TraitPair, DeriveJunction, SecretStringError, Ss58Co
 #[cfg(feature = "std")]
 use serde::{de, Serializer, Serialize, Deserializer, Deserialize};
 use crate::{crypto::{Public as TraitPublic, UncheckedFrom, CryptoType, Derive}};
+use core::{str::FromStr, ops::Deref};
```
Comment (bkchr) [inline core/primitives/src/ed25519.rs]:
```suggestion
use rstd::ops::Deref;
```

File: core/primitives/src/ed25519.rs [hunk 2]
```diff
@@ -111,6 +105,15 @@ impl From<Public> for H256 {
 	}
 }
 
+#[cfg(feature = "std")]
+impl FromStr for Public {
```
Comment (bkchr) [inline core/primitives/src/ed25519.rs]:
```suggestion
impl std::str::FromStr for Public {
```

File: core/primitives/src/sr25519.rs [hunk 3]
```diff
@@ -31,11 +31,12 @@ use substrate_bip39::mini_secret_from_entropy;
 use bip39::{Mnemonic, Language, MnemonicType};
 #[cfg(feature = "std")]
 use crate::crypto::{
-	Pair as TraitPair, DeriveJunction, Infallible, SecretStringError, Ss58Codec
+	Pair as TraitPair, DeriveJunction, Infallible, SecretStringError, Ss58Codec, PublicError
 };
 use crate::{crypto::{Public as TraitPublic, UncheckedFrom, CryptoType, Derive}};
 use crate::hash::{H256, H512};
 use codec::{Encode, Decode};
+use core::{str::FromStr, ops::Deref};
```
Comment (bkchr) [inline core/primitives/src/sr25519.rs]:
```suggestion
use rstd::ops::Deref;
```

File: core/primitives/src/sr25519.rs [hunk 4]
```diff
@@ -96,6 +90,15 @@ impl From<Public> for H256 {
 	}
 }
 
+#[cfg(feature = "std")]
+impl FromStr for Public {
```
Comment (bkchr) [inline core/primitives/src/sr25519.rs]:
```suggestion
impl std::str::FromStr for Public {
```

**Comment by c410-f3r** [comment]
It is now making use of the reexported `FromStr` of `rstd`. Is it OK?

**Comment by bkchr** [comment]
Could you please use all my suggestions. Otherwise it will generate warnings for the wasm build.

**Comment by c410-f3r** [comment]
Updated

**Comment by gavofyork** [comment]
`Public` is not `Copy` for a reason.

**Comment by bkchr** [comment]
> `Public` is not `Copy` for a reason.

What is the reason?

**Comment by gavofyork** [comment]
To make the clones operation explicit and avoid accidentally pessimising.

**Comment by bkchr** [comment]
I think everything up to a cache line should be safe to implement `Copy`. Besides that, we already have in the code base:
1. `u128` and that also implements `Copy`.
2. Stuff like [`H512`](https://docs.rs/primitive-types/0.6.1/primitive_types/struct.H512.html) also implements `Copy`.

**Comment by c410-f3r** [comment]
IIRC, `clone` on `[T; N]` (T: Copy) is a simply `memcpy`, threfore, pretty much the same thing as `copy`. Even though (ed25519|sr25519)::Public isn't `repr(transparent)` and "copiable", the compiler probably already optimizes `clone` to `memcpy` the inner array.

Personally, if you guys are willing to merge this PR without the `Copy` implementation, I will gladly remove it.

**Comment by gavofyork** [comment]
Ok - I've no strong opinions on the matter so happy to approve. @bkchr just waiting for an approval from you.

**Comment by c410-f3r** [comment]
Squashed and rebased

**Comment by c410-f3r** [comment]
Thanks!

**Pull request merged**